Add new LocalFilesystemToS3Operator under Amazon provider (#17168)

denimalpaca opened the pull request:
Add new LocalFilesystemToS3Operator under Amazon provider (#17168)

Currently, an S3Hook exists that allows transfer of files to S3 via `load_file()`, however there is no operator associated with it. The LocalFilesystemToS3Operator would wrap the S3 Hook, so it is not used directly.

Seeing as to upload a local file to S3 using the S3 Hook, a Python task with the same functionality has to be written anyway, this could reduce a lot of redundant boiler-plate code and standardize the local file to S3 load process.

closes: #17168

Commit:
Add new LocalFilesystemToS3Operator under Amazon provider (#17168)

Fix `airflow celery stop` to accept the pid file. (#17278)

Fix isort (#17330)

Fix failing celery test (#17337)

This change fixes failing test due to mocking

subkanthi commented on airflow/providers/amazon/aws/example_dags/example_local_to_s3.py:
```diff
@@ -0,0 +1,40 @@
+    start_date=datetime(2021, 1, 1),  # Override to match your needs
+) as dag:
+    # [START howto_local_transfer_data_to_s3]
+    create_sftp_to_s3_job = LocalFilesystemToS3Operator(
```
maybe a copy/paste, create_sftp_to_s3_job should be I guess local_to_s3

denimalpaca replied:
d'oh. Yes. Thank you.

denimalpaca committed:
Fix the task variable name and add task instance to DAG

denimalpaca committed:
Merge branch 'apache:main' into main

denimalpaca committed:
Merge branch 'apache:main' into main

denimalpaca committed:
Merge branch 'apache:main' into main

ferruzzi commented on airflow/providers/amazon/aws/example_dags/example_local_to_s3.py:
```diff
@@ -0,0 +1,42 @@
+from airflow.providers.amazon.aws.transfers.local_to_s3 import LocalFilesystemToS3Operator
+from airflow.utils.dates import datetime
+
+S3_BUCKET = os.environ.get("S3_BUCKET", "test-bucket")
```
Nitpick for discussion, non-blocking:  There are currently four example DAGS (not counting this one) related to S3.  Two of them use the envvar "S3_BUCKET" and the other two use "BUCKET_NAME".  Should we standardize those and, if so, which way?

denimalpaca replied:
I would opt for "S3_BUCKET" as the hook uses `dest_bucket`, and it seems that "S3_BUCKET" over "BUCKET_NAME" more closely follows that precedent.

mik-laj replied:
Let's do this as a separate contribution because it makes sense to standardize all the variables and write it down as a generally accepted rule to make it more permanent. Now, if we change the variable name in one place, we will fix the problem only here, but in a moment subsequent contributions may not see this convention and introduce a different naming convention.

ferruzzi commented on airflow/providers/amazon/aws/transfers/local_to_s3.py:
```diff
@@ -0,0 +1,107 @@
+    :type dest_key: str
+    :param dest_bucket: Name of the S3 bucket to where the object is copied. (templated)
+
+        It should be omitted when `dest_key` is provided as a full s3:// url.
```
Nit:  I'd really prefer to see some logic to enforce this rather than a comment.

uranusjr replied:
+1 but I don’t think this is a nit, we should enforce this in code (e.g. raise TypeError if both are passed) and describe it here.

denimalpaca committed:
Add test for operator init exception

When dest_key is given as a full s3:// filepath and dest_bucket is
also provided, operator throws an error in its init() method. A test
is added for this case.

denimalpaca committed:
Add Exception to Operator init when mutually exclusive parameters are given

When dest_key is given as a full s3:// file path, the parent operator does
not expect a dest_bucket. A TypeError is raised when dest_bucket is not None
in this case.

mik-laj reviewed (commented):
LGTM, But CI hasn't been running for this change yet
